README.txt
Introduction - what is XInstaReap?
::::::::::::::::::::::::::::::::::

XInstaReap is a multiplayer mod for the 6DOF game "Descent 3" ((C) Outrage
Entertainment) which extends the original mod "InstaReap" by "Grim" in which
you only have the mass driver weapon which will kill your enemy players with
just one hit.


What are the differences between XInstaReap and InstaReap?
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::

XInstaReap currently has the following changes (as of version 23):

Bugfixes:
	-When shooting the first time, the weapon changes to the mass driver
         and shoots immediately instead of waiting for another shot (the
         mass driver can't be equipped right after spawning because of a
         bug in the reticle management)

New features:
	-Free afterburner coolers for everyone!
	-Cheat prevention on server side by measuring the time between two
	 shots per player - cheaters will be kicked and all players will be
	 notified (by using text and the classic "Cheater!" sound)
	-All shots are now displayed in the color of the player who is shooting

Removed stuff:
	-The Outrage logo won't be displayed anymore

New restrictions:
	-XInstaReap only supports up to 16 players because of not having enough
	 applicable weapons for coloring shots (starting a server configured
	 for more than 16 players will automatically set max players to 16)
	-You should use a dedicated server for this mod to work probably. If
         you start a server ingame, some features like cheat protection might
         not work properly


Installation
::::::::::::

Linux
-----

-Extract the files XInstaReap.d3m, XInstaReap30.d3m and XInstaReap_o.so to
 ~/.loki/descent3/netgames/
-Copy the library file dmfc.so to /usr/local/lib


Windows
-------

-Extract the files XInstaReap.d3m, XInstaReap30.d3m and XInstaReap_o.dll to
 <Your Descent3 installation>\netgames\


Source code
:::::::::::

The code of XInstaReap is available to anyone on its project page (see bottom).


Compiling
:::::::::

The mod can be compiled by any supported compiler of the Descent 3 SDK. The mod
has been built on a Linux machine with GCC 2.95-3 and MSVC 6.

On Linux, type "make" (for debug) or "make release" (for optimized binary).
On Windows, load the project anarchy.dsp and select your target (debug or
release).


Q & A
:::::

Q: Why are the sources named "Anarchy.cpp" instead of "XInstaReap.cpp", etc.?
A: That's because I was too lazy to rename the project in MSVC so if you build
   the mod on Windows, you will also get a file "anarchy.dll" which you need to
   rename to XInstaReap.dll. The Linux Makefile automatically builds the file
   XInstaReap.so.

Q: What can I do with this source code?
A: Anything you want, except for selling it or binaries of it (see below).

Q: What is the license of the mod?
A: Since XInstaReap and InstaReap were built from the original mod "Anarchy" by
   Outrage, this mod uses the Outrage License which can be found in the header
   of each source code file or in the file LICENSE.txt.

Q: If I want to modify this mod, where can I get the compilers?
A: As for MSVC 6, you will need to buy a copy from eBay. As for GCC on Linux,
   you can either compile an old version (<= 2.95-3) or ask D.Cent for a
   portable chroot (see below).


Credits
:::::::

-Daniel ("Grim") - original version of InstaReap - big thanks to you!
-Philipp Lorenz ("VEX-D.Cent") - changes for XInstaReap
-VEX clan - thank you for testing :)


Contact
:::::::

You can either contact me on the project page of XInstaReap or on IRC:

Project page: https://www.github.com/WindohsCrasher/XInstaReap
IRC: irc.descentforum.net, channel "#Descent", player "VEX-D_Cent"


ChangeLog.txt
Revision 25
===========

-Efficiency is now displayed with a decimal place of 2


Revision 24
===========

-Fix for black pyro's flares kicking the player because they are considered a
 different weapon
-Added shoot tolerance of 0.1s for anti cheat


Revision 23
===========

-Fixed weapon reload time not restoring after first shot
-Added "cheater" sound notification if cheats were detected


Revision 22
===========

-Fixed behavior of switching weapons after respawning blocking the first shot
-Modified cheat message so people who experience a lot of lag (which is
 definitely not cheating) will not be blamed to be cheaters


Revision 21
===========

-First public source code release


Anarchy.h
/*
=========================================================
THE COMPUTER CODE CONTAINED HEREIN IS THE SOLE 
PROPERTY OF OUTRAGE ENTERTAINMENT, INC. 
('OUTRAGE').  OUTRAGE, IN DISTRIBUTING THE CODE TO
END-USERS, AND SUBJECT TO ALL OF THE TERMS AND 
CONDITIONS HEREIN, GRANTS A ROYALTY-FREE, 
PERPETUAL LICENSE TO SUCH END-USERS FOR USE BY 
SUCH END-USERS IN USING, DISPLAYING,  AND 
CREATING DERIVATIVE WORKS THEREOF, SO LONG AS
SUCH USE, DISPLAY OR CREATION IS FOR NON-
COMMERCIAL, ROYALTY OR REVENUE FREE PURPOSES. 
IN NO EVENT SHALL THE END-USER USE THE 
COMPUTER CODE CONTAINED HEREIN FOR REVENUE-
BEARING PURPOSES.  THE END-USER UNDERSTANDS
AND AGREES TO THE TERMS HEREIN AND ACCEPTS THE 
SAME BY USE OF THIS FILE.
COPYRIGHT 1999 OUTRAGE ENTERTAINMENT, INC.  ALL 
RIGHTS RESERVED.
=========================================================
*/

/*
* $Logfile: /DescentIII/Main/anarchy/Anarchy.h $
* $Revision: 32 $
* $Date: 5/12/99 11:28a $
* $Author: Jeff $
*
* <insert description of file here>
*
* $Log: /DescentIII/Main/anarchy/Anarchy.h $
 * 
 * 32    5/12/99 11:28a Jeff
 * added sourcesafe comment block
*
* $NoKeywords: $
*/



#ifndef __DMFC_APP_H_
#define __DMFC_APP_H_

#include "osiris_share.h"
#include "d3events.h"

//Setup and processing functions
void AnarchyGameInit(int teams);
void AnarchyGameClose(void);
void SendChangeWeaponsRequest(int iPnum);
void ReceiveSendChangeWeaponsRequest(ubyte *uData);
void ReceiveChangeWeapons(ubyte *uData);
void CheckVersionNumber(ubyte *data);
void GetVersionNumber(int pnum);
void AddScoreToPlayer(ubyte *data);
void OnServerPlayerDisconnect(int player_num);
void OnServerPlayerEntersGame(int player_num);
void CheckPlayers(int number);
void OnServerCollide(object *me_obj,object *it_obj);
int CheckWeapon(int slot, int weap_index);
int  ChangeWeapon(int slot, int weap_index);
int  RemWeaponFromPlayer(int slot, int weap_index);
int  AddWeaponToPlayer(int slot,int weap_index,int ammo); 
void OnKeypress(int key);
void OnHUDInterval(void);
void OnInterval(void);
void OnClientPlayerKilled(object *killer_obj,int victim_pnum);
void OnServerPlayerKilled(object *killer_obj,int victim_pnum);
void OnClientPlayerEntersGame(int player_num);
void OnClientLevelStart(void);
void OnClientLevelEnd(void);
void OnServerGameCreated(void);
void OnPLRInterval(void);
void OnPLRInit(void);
void OnSaveStatsToFile(void);
void OnLevelEndSaveStatsToFile(void);
void OnDisconnectSaveStatsToFile(void);
void OnPrintScores(int level);
extern IDMFC *DMFCBase;



/*************************************************************************************************
 *The following functions and declaration are needed to connect the DLL to the game.  These must *
 *stay here and must call the functions that are in them.  You can not delete from here, but you *
 *can add to it no problem																								 *
 *************************************************************************************************
*/
// These next two function prototypes MUST appear in the extern "C" block if called
// from a CPP file.
#ifdef __cplusplus
extern "C"
{
#endif
	DLLEXPORT void DLLFUNCCALL DLLGameInit (int *api_func,ubyte *all_ok,int num_teams_to_use);
	DLLEXPORT void DLLFUNCCALL DLLGameCall (int eventnum,dllinfo *data);
	DLLEXPORT void DLLFUNCCALL DLLGameClose ();
	DLLEXPORT void DLLFUNCCALL DLLGetGameInfo (tDLLOptions *options);
	DLLEXPORT int DLLFUNCCALL GetGOScriptID(char *name,ubyte isdoor);
	DLLEXPORT void DLLFUNCCALLPTR CreateInstance(int id);
	DLLEXPORT void DLLFUNCCALL DestroyInstance(int id,void *ptr);
	DLLEXPORT short DLLFUNCCALL CallInstanceEvent(int id,void *ptr,int event,tOSIRISEventInfo *data);
	DLLEXPORT int DLLFUNCCALL SaveRestoreState( void *file_ptr, ubyte saving_state );
#ifdef __cplusplus
}
#endif


#endif


anarchystr.h
/*
=========================================================
THE COMPUTER CODE CONTAINED HEREIN IS THE SOLE 
PROPERTY OF OUTRAGE ENTERTAINMENT, INC. 
('OUTRAGE').  OUTRAGE, IN DISTRIBUTING THE CODE TO
END-USERS, AND SUBJECT TO ALL OF THE TERMS AND 
CONDITIONS HEREIN, GRANTS A ROYALTY-FREE, 
PERPETUAL LICENSE TO SUCH END-USERS FOR USE BY 
SUCH END-USERS IN USING, DISPLAYING,  AND 
CREATING DERIVATIVE WORKS THEREOF, SO LONG AS
SUCH USE, DISPLAY OR CREATION IS FOR NON-
COMMERCIAL, ROYALTY OR REVENUE FREE PURPOSES. 
IN NO EVENT SHALL THE END-USER USE THE 
COMPUTER CODE CONTAINED HEREIN FOR REVENUE-
BEARING PURPOSES.  THE END-USER UNDERSTANDS
AND AGREES TO THE TERMS HEREIN AND ACCEPTS THE 
SAME BY USE OF THIS FILE.
COPYRIGHT 1999 OUTRAGE ENTERTAINMENT, INC.  ALL 
RIGHTS RESERVED.
=========================================================
*/

#ifndef __ANARCHY_STRING_TABLE_H_
#define __ANARCHY_STRING_TABLE_H_

#define TXT(d)	GetString(d)

#define TXT_DEATH1		TXT(0)	//"%s got blasted by %s"
#define TXT_DEATH2		TXT(1)	//"%s knows %s is his god"
#define TXT_DEATH3		TXT(2)	//"%s sucks %s's milk"
#define TXT_DEATH4		TXT(3)	//"%s realizes %s's power"
#define TXT_DEATH5		TXT(4)	//"%s got killed by %s"
#define TXT_DEATH6		TXT(5)	//"%s begs for %s's mercy"
#define TXT_DEATH7		TXT(6)	//"%s realizes %s is a better player"
#define TXT_DEATH8		TXT(7)	//"%s was no match for %s"
#define TXT_DEATH9		TXT(8)	//"%s wishes he was as good as %s"
#define TXT_DEATH10		TXT(9)	//"%s got messed up by %s"
#define TXT_SUICIDE1	TXT(10)	//"%s blasts himself"
#define TXT_SUICIDE2	TXT(11)	//"%s Bursts his own bubble"
#define TXT_SUICIDE3	TXT(12)	//"%s doesn't know his own strength"
#define TXT_SUICIDE4	TXT(13)	//"No prize for %s"
#define TXT_SUICIDE5	TXT(14)	//"%s doesn't wish to live anymore"
#define TXT_SUICIDE6	TXT(15)	//"%s SUCKS!"
#define TXT_PILOT		TXT(16)	//"Pilot"
#define TXT_KILLS		TXT(17)	//"Kills"
#define TXT_DEATHS		TXT(18)	//"Deaths"
#define TXT_SUICIDES	TXT(19)	//"Suicides"
#define TXT_SCORE		TXT(20)	//"Score"
#define TXT_STATS		TXT(21)	//"Stats"
#define TXT_WELCOME		TXT(22)	//"Welcome to the Anarchy %s!"
#define TXT_JOINED		TXT(23)	//"%s has joined the Anarchy"
#define TXT_POINTS		TXT(24)	//"Points"
#define TXT_HUDD_NONE	TXT(25)	//"Anarchy: HUD Display set to None"
#define TXT_HUDD_SCORES	TXT(26)	//"Anarchy: HUD Display set to Scores"
#define TXT_HUDD_EFFIC	TXT(27)	//"Anarchy: HUD Display set to Efficiency"
#define TXT_MNU_HUDSTYLE	TXT(28)	//"HUD Display"
#define TXT_NONE			TXT(29)	//"None"
#define TXT_SCOREHD			TXT(30)	//"Score"
#define TXT_EFFICIENCY		TXT(31)	//"Efficiency"
#define TXT_MNU_HUDCOLOR	TXT(32)	//"HUD Score Colors"
#define TXT_PLAYERCOLORS	TXT(33)	//"Player Colors"
#define TXT_NORMAL			TXT(34)	//"Normal"
#define TXT_PING			TXT(35)	//"Ping"
#define TXT_SAVE_HEADER		TXT(36)	//"Anarchy\r\nGame: %s\r\nLevel: %d\r\n"
#define TXT_SAVE_HEADERB	TXT(37)	//"[Rank] [Name]                      [Score]     [Kills]     [Deaths]   [Suicides] [Time In Game]"
#define TXT_SAVE_HEADERC	TXT(38)	//"\r\nIndividual Stats\r\n"
#define TXT_SAVE_TIMEINGAME	TXT(39)	//"Total Time In Game: %s"
#define TXT_SAVE_KILLERLIST	TXT(40)	//"Callsign:                     Kills:    Deaths:"
#define TXT_MSG_SAVESTATS	TXT(41)	//"Stats saved to file"
#define TXT_MSG_COLORPLR	TXT(42)	//"Anarchy HUD Color Model: Player Colors"
#define TXT_MSG_COLORNORM	TXT(43)	//"Anarchy HUD Color Model: Normal"
#define TXT_KILLS_SHORT		TXT(44)	//"K"
#define TXT_DEATHS_SHORT	TXT(45)	//"D"
#define TXT_SUICIDES_SHORT	TXT(46)	//"S"

#endif
